docs/design/git-submodule-storage.md

```diff
@@ -0,0 +1,146 @@
+
+## Use cases to consider
+
+The submodule storage format should support the workflows specified in the
+[submodules roadmap](./git-submodules.md). It should be obvious how "Phase 1"
+requirements will be supported, and we should have an idea of how "Phases 2,3,X"
+might be supported.
+
+Notable use cases and workflows are noted below.
+
+### Fetching submodule commits
+
+Git's protocol is designed for communicating between copies of the same
+repository. Notably, a Git fetch calculates the list of required objects by
+performing reachability checks between the refs on the local and the remote
+side. We should expect that this will only work well if the submodule repository
+is stored as a local Git repository.
+
+Rolling our own Git fetch is too complex to be worth the effort.
+
+### "jj op restore" and operation log format
+
+We want `jj op restore` to restore to an "expected" state in the submodule.
+There is a potential distinction between running `jj op restore` in the
+superproject vs in the submodule, and the expected behavior may be different in
+each case, e.g. in the superproject, it might be enough to restore the submodule
+working copy, but in the submodule, refs also need to be restored.
+
+Currently, the operation log only references objects and refs in the
+superproject, so it is likely that proposed approaches will need to extend this
+format. It is also worth considering that submodules may be added, updated or
+removed in superproject commits, thus the list of submodules is likely to change
+over the repository's lifetime.
+
+### Nested submodules
+
+Git submodules may contain submodules themselves, so our chosen storage schemes
+should support that.
+
+We should consider limiting the recursion depth to avoid nasty edge cases (e.g.
+cyclical submodules.) that might surprise users.
+
+### Supporting future extensions
+
+There are certain extensions we may want to make in the future, but we don't
+have a timeline for them today. Proposed approaches should take these
+extensions into account (e.g. the approach should be theoretically extensible),
+but a full proposal for implementing them is not necessary.
+
+These extensions are:
+
+- Non-git subrepos
+- Colocated Git repos
+- Non-git backends
+
+## Proposed design
+
+Git submodules will be stored as full jj repos. In the code, jj commands will
+only interact with the submodule's repo as an entire unit, e.g. it cannot query
+the submodule's commit backend directly. A well-abstracted submodule will extend
+well to non-git backends and non-git subrepos.
+
+The main challenge with this approach is that the submodule repo can be in a
+state that is internally valid (when consider only the submodule's repo), but
+invalid when considering the superproject-submodule system. This will be managed
+by requiring all submodule interactions go through the superproject so that
+superproject-submodle coordination can occur. For example, jj will not allow the
```

Comment:
You left some typos `(when considering only the submodule's repo)` and `superproject-submodule`

Reply:
Ah, thanks!